grader_systems/src/stage/valgrind.rs

```diff
@@ -0,0 +1,1123 @@
+            return Err(anyhow!(
+                "Found empty valgrind log at {:?}. Something went wrong.",
+                path.to_str()
+            ));
+        }
+
+        // replace all absolute paths with basename
+        let re = Regex::new(r"(\W|^)(?:\/[^\/\s]+)+\/([^\/\s]+)\b")?;
+        let repl = re.replace_all(&contents, "$1$2");
+
+        Ok(repl.to_string())
+    }
+}
+
+#[async_trait]
+impl<E: ProcessExecutor> Executor for Valgrind<E> {
+    type Output = StageResult;
+
+    async fn run(&self, ws: &Path) -> Result<Self::Output> {
+        let mut sect = Section::new("Valgrind");
+        let mut results_sect = Section::new("Output:");
+        let mut run_updates = StatusUpdates::default();
+        debug!("running program");
+
+        let executable = ws.join(&self.executable);
+        if !executable.exists() {
+            return Err(anyhow!(
+                "Could not find student executable at {:?}",
+                executable.display()
+            ));
+        }
+
+        if let Some(v) = &self.config.suppressions {
+            for supp in v {
+                if !ws.join(&supp).exists() {
+                    return Err(anyhow!("Could not find valgrind suppressions at {}", supp));
+                }
+            }
+        }
+
+        let cmd = self.gen_cmd(ws);
+        sect.add_content(("Run Command", format!("{}", cmd).code()));
+
+        let res = cmd.run_with(&self.executor).await?;
+
+        let log = self.read_logfile(ws)?;
+
+        if !res.status.completed() {
+            match res.status {
+                ExitStatus::Timeout(to) => {
+                    run_updates.add_update(
+                        Update::new_fail("running valgrind", PointQuantity::FullPoints).notes(
+                            format!(
+                                "Your submission timed out after {} seconds :(",
+                                to.as_secs()
+                            ),
+                        ),
+                    );
+                    sect.add_content(run_updates);
+                }
+                ExitStatus::Signal(sig) => match sig {
+                    SignalType::SegFault => {
+                        run_updates.add_update(
+                            Update::new_fail("running valgrind", PointQuantity::FullPoints)
+                                .notes("Your submission was killed by SIGSEGV!"),
+                        );
+                        results_sect.add_content(log);
+                        sect.add_content(run_updates);
+                        sect.add_content(Content::SubSection(results_sect));
+                    }
+                    SignalType::Abort => {
+                        run_updates.add_update(
+                            Update::new_fail("running valgrind", PointQuantity::FullPoints)
+                                .notes("Your submission was killed by SIGABRT!"),
+                        );
+                        results_sect.add_content(log);
```

Comment:
You do this in almost every branch, why not just add the log after you create it instead of doing it so many different places?

Reply:
Yeah the output is tricky because it needs to be invoked in the order it should appear, and the timeout should not print the log. I considered compressing both signals into one block and using an if statement for the signal type, but chose the match expression for the static assertion. But I should just do that and make an empty match expression below it to avoid code duplication

Reply:
Hmm, yeah I see your point. Ironically, generating good output was always one of the hardest parts of the autograder. if you have any ideas to make it more ergonomic, let me know. 

What you said makes sense. In some ways, it could be less confusing to have that nice parallel between branches of adding the seciton in, especially if only one branch doesn't need it. I'll leave it up to you to decide which method looks the neatest.